Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -0,0 +1,449 @@
+        /// <summary>
+        /// List of known orders
+        /// </summary>
+        public ConcurrentDictionary<int, Order> CachedOrderIDs = new ConcurrentDictionary<int, Order>();
+        /// <summary>
+        /// List of filled orders
+        /// </summary>
+        protected readonly FixedSizeHashQueue<int> FilledOrderIDs = new FixedSizeHashQueue<int>(10000);
+        /// <summary>
+        /// List of unknown orders
+        /// </summary>
+        protected readonly FixedSizeHashQueue<int> UnknownOrderIDs = new FixedSizeHashQueue<int>(1000);
+        /// <summary>
+        /// Name of wallet
+        /// </summary>
+        protected string Wallet;
+        const string _exchange = "bitfinex";
+        /// <summary>
+        /// Api Key
+        /// </summary>
+        protected string ApiKey;
+        /// <summary>
+        /// Api Secret
+        /// </summary>
+        protected string ApiSecret;
+        TradingApi.Bitfinex.BitfinexApi _restClient;
+        /// <summary>
+        /// Security Provider
+        /// </summary>
+        protected ISecurityProvider SecurityProvider;
+        #endregion
+
+        /// <summary>
+        /// Create bitfinex brokerage
+        /// </summary>
+        public BitfinexBrokerage(string apiKey, string apiSecret, string wallet, BitfinexApi restClient, decimal scaleFactor, ISecurityProvider securityProvider)
+            : base("bitfinex")
+        {
+            ApiKey = apiKey;
+            ApiSecret = apiSecret;
+            Wallet = wallet;
+            _restClient = restClient;
+            ScaleFactor = scaleFactor;
+            SecurityProvider = securityProvider;
+        }
+
+        /// <summary>
+        /// Determines if ticker polling is active
+        /// </summary>
+        public override bool IsConnected
+        {
+            get { return this._tickerToken != null && !this._tickerToken.IsCancellationRequested; }
+        }
+
+        private decimal GetPrice(Order order)
+        {
+            if (order is StopMarketOrder)
+            {
+                return ((StopMarketOrder)order).StopPrice * ScaleFactor;
+            }
+            else if (order is LimitOrder)
+            {
+                return ((LimitOrder)order).LimitPrice * ScaleFactor;
+            }
+
+            return order.Price <= 0 ? order.Id : (order.Price * ScaleFactor);
+        }
+
+        /// <summary>
+        /// Place a new order
+        /// </summary>
+        /// <param name="order"></param>
+        /// <returns></returns>
+        public override bool PlaceOrder(Orders.Order order)
+        {
+            //todo: wait for callback from auth before posting
+            Authenticate();
+
+            int quantity = (int)Math.Floor(SecurityProvider.GetHoldingsQuantity(order.Symbol));
+            Orders.Order crossOrder = null;
+            if (OrderCrossesZero(order, quantity))
+            {
+                crossOrder = order.Clone();
+                var firstOrderQuantity = -quantity;
```

Comment:
I'm a little confused as to why we're doing the zero crossing logic here, but then in the `private bool PlaceOrder` below we immediately recombine them into a single order.  Some brokerages require orders to be split up and some even require the order that brings the holdings to zero to fill before submitting the next order, but it doesn't look like that's required here.. or is it?